### ISSUE TITLE
Unique constraints don't work as expected

### ISSUE THREAD

**Issue by jpmckinney** [open-contracting/kingfisher-process#235]
If a unique constraints contains a nullable field, then there can be multiple rows with 'identical' values, because NULL != NULL. This occurs with:

1. unique_collection_identifiers: source_id, data_version, sample, transform_from_collection_id (nullable), transform_type (nullable)
1. unique_collection_file_identifiers: collection, filename (nullable)
1. unique_collection_file_item_identifiers: collection_file, number (nullable)
1. unique_release_check_release_id_and_more: release, override_schema_version (nullable)
1. unique_record_check_record_id_and_more: record, override_schema_version (nullable)
1. unique_release_check_error_release_id_and_more: release, override_schema_version (nullable)
1. unique_record_check_error_record_id_and_more: record, override_schema_version (nullable)

**Separately:** Why is there unique_collection_file_identifiers for filename, but not another index for url?

For 2 and 4-7, the solution is easy: Don't allow NULL text fields, which is a [best practice](https://docs.djangoproject.com/en/3.0/ref/models/fields/#null): there should only be one 'no data' value (the empty string). So, a good first PR would be to make these non-nullable.

For 3, under what condition is `number` NULL? If there is no such condition, we can set it to NOT NULL.

For 1, I think we first need to reiterate what this constraint is actually meant to do, before considering how to fix it.

* We want to allow multiple collections of the same source at different times, so we need at least `source_id` and `data_version` in a unique constraint.
* I don't know why `sample` is participating in the constraint. If an analyst ran a sample collection and then a full collection from scrapy, wouldn't the `data_version` always differ?
* We don't want many of the same transformation of the same source collection. For that, we just need a unique constraint on `transform_from_collection_id` and `transform_type`. Otherwise, we can run the same transform on the same collection multiple times by changing `source_id` or `data_version`. The only condition where I think we'd want to do that is if the transform fails. However, the protocol should be to delete (or restart) that transformation, rather than grow the number of collections (for example, if an analyst/developer makes a change, tries again, fails, makes a change, tries again, etc.).

So, in terms of a solution, I think we need:

* A unique constraint on `source_id`, `data_version`, `transform_type`. That way, we have only one collection for a given source and time, and only one of each transform of that collection.
* While the above constraint would suffice, I think we'd want another constraint (or at least an index to support the query that checks for an existing destination collection) on `transform_from_collection_id` and `transform_type`, since `transform_from_collection_id` is a more robust way of identifying a source collection than `source_id`, `data_version` and `transform_type` together.

Otherwise, to keep the current unique constraint, we'll need a [partial unique index](https://dba.stackexchange.com/a/9760) that excludes `transform_from_collection_id` where it is NULL. I prefer the above option.

If we want to optimize the constraints, we could have a partial unique index on `source_id` and `data_version` where `transform_type` is empty (i.e. one collection for a given source and time), and another partial unique index on `transform_from_collection_id` and `transform_type` where `transform_from_collection_id` is not NULL (i.e. one transform of a given type per collection).

**Comment by jpmckinney** [comment]
```sql
SELECT source_id, data_version, transform_type, COUNT(*) FROM collection GROUP BY source_id, data_version, transform_type HAVING COUNT(*) > 1;
```

returns:

```
               source_id               |    data_version     |  transform_type  | count 
---------------------------------------+---------------------+------------------+-------
 buenos-aires                          | 2019-02-21 10:00:12 | compile-releases |     2
 chile_compra_records                  | 2019-12-03 10:34:55 | compile-releases |     2
 uruguay_historical                    | 2019-06-13 12:59:56 | compile-releases |     2
 mexico_administracion_publica_federal | 2019-03-18 14:44:32 | compile-releases |     2
```

So, there are cases of creating multiple transforms of the same type for the same collection.

**Comment by jpmckinney** [comment]
> I don't know why sample is participating in the constraint. If an analyst ran a sample collection and then a full collection from scrapy, wouldn't the data_version always differ?

I suppose if the analyst is local-loading a file (setting `data_version` to the time they e.g. downloaded the bulk file) first as a sample and then in full, they would set the `data_version` to the same value both times (as its semantics are to describe the time at which the data was collected, not the time at which it was loaded), and so that's the reason `sample` is in the index. If that's correct, then we need to include it in the first proposed new index.

**Comment by odscjames** [comment]
> because NULL != NULL

So I'll hold my hand up here, I've just been looking at NULL values in NULL columns in UNIQUE CONSTRAINTS across multiple columns and how they are enforced and my understanding was wrong. This does bring up some issues, but there is a lot in this thread, so I'll reply soon when I've had time to look at this properly.

**Comment by odscjames** [comment]
>    unique_collection_identifiers: source_id, data_version, sample, transform_from_collection_id (nullable), transform_type (nullable)

One motiviation at the time of setting this up was that the transforms section was not fully specced, and so it was designed to be a bit lose. If we are now clear about what transforms we will have, we can tighten this up a bit.

Something like you suggest could work well, tho I note that in https://github.com/open-contracting/kingfisher-process/issues/240#issuecomment-566111117 you talk about new transforms or options to transforms and that should be double checked. If there was an original data set 1, and an analyst wanted data set 2 with transform EXAMPLE with some options, and also data set 3 with transform EXAMPLE with some different options, we need to make sure that is supported.

>We want to allow multiple collections of the same source at different times, so we need at least source_id and data_version in a unique constraint.
>I suppose if an analyst downloaded a bulk file and is local-loaded it first as a sample and then in full, 

I would include sample; I don't know any harm it does and it makes it clearer and simple to operaters what is and isn't allowed in terms of duplicates. There is another reason to include sample around Scrape which I'll get into below.

>A unique constraint on source_id, data_version, transform_type. That way, we have only one collection for a given source and time, and only one of each transform of that collection.
> ..... on transform_from_collection_id and transform_type

This has issues with transform options? But we can probably work out a way to work them in.

The other option is to relax these unique constraints a bit around transforms. 

The first system allowed the identification of specific  data sets by a set of variables - it had to, because it used file storage! The second system (The Scrape-Process split) moved to a system where every collection in Process was in the database and had it's own integer ID. So the question is, do we need any unique constraints at all?

There is one reason - Scrape sends data to Process with 3 flags ("collection_source", "collection_data_version", "collection_sample") and Process must identify the correct ID of collection  to put the data into. So we somehow need a unique index on those 3 flags for collections that are not transforms.

But, maybe we can do away with unique indexes on stuff around transforms all together? The system could warn operators if they tried to create a duplicate transform, but ultimately it would let them if they really wanted.




>    unique_collection_file_identifiers: collection, filename (nullable)

I think the solution here is that filename should not be nullable anyway? I can't think of a use case where this is wanted. I'm hedging because I can't remember why we did this originally, so I'm wary I've just forgotten the reason (or the reason was historical, or it was an oversight).

In fact, when the store file Item API is used the system uses filename as a unique ID, so it needs a good name anyway. (This API is typically only used for JSON Lines data.)

	ocdskingfisherprocess=> select * from collection_file where filename is null;
	    id    | collection_id | filename | store_start_at | store_end_at | warnings |                                   url                                   |       errors       
	----------+---------------+----------+----------------+--------------+----------+-------------------------------------------------------------------------+--------------------
	 88699416 |           813 |          |                |              |          | https://apis.mercadopublico.cl/OCDS/data/listaA%C3%B1oMes/2008/01/0/100 | {"http_code": 404}
	(1 row)

If we add an kf_filename to https://github.com/open-contracting/kingfisher-scrape/blob/master/kingfisher_scrapy/spiders/chile_base.py#L50 I think that will stop that happening again.
This is easy to miss in scrapers, especially as this arises around errors only; maybe Process could assign a random filename if a blank one is passed?

>    unique_collection_file_item_identifiers: collection_file, number (nullable)

I think the solution here is that number should not be nullable anyway? Again, I'm hedging because I can't remember why we did this originally, so I'm wary I've just forgotten the reason (or the reason was historical, or it was an oversight).

	ocdskingfisherprocess=> select * from collection_file_item where number is null;
	 id | collection_file_id | store_start_at | store_end_at | number | warnings | errors 
	----+--------------------+----------------+--------------+--------+----------+--------
	(0 rows)

>    unique_release_check_release_id_and_more: release, override_schema_version (nullable)
>    unique_record_check_record_id_and_more: record, override_schema_version (nullable)
>    unique_release_check_error_release_id_and_more: release, override_schema_version (nullable)
>    unique_record_check_error_record_id_and_more: record, override_schema_version (nullable)

override_schema_version can be made NOT NULL, and we can use an empty string '' to indicate that no override was set.

**Comment by jpmckinney** [comment]
> Why is there unique_collection_file_identifiers for filename, but not another index for url?

1. We don't need it. Filename is the main thing to make unique.
2. The collection_file for a compiled release doesn't set url (and filename is set to `<ocid>.json`).

**Comment by jpmckinney** [comment]
I made a first PR for the easier ones: #249

> maybe we can do away with unique indexes on stuff around transforms all together? The system could warn operators if they tried to create a duplicate transform, but ultimately it would let them if they really wanted.

That's fine with me. I'll make a separate PR for it. 

> maybe Process could assign a random filename if a blank one is passed?

I think Process should return an error if no filename is passed. I don't know if Scrapy has any validation hooks we can use, so that spider authors can get an error during development.

**Comment by jpmckinney** [comment]
I had to leave setting release_check.override_schema_version to `''` out of #249, because:

```sql
SELECT COUNT(*) FROM (
    SELECT release_id, COUNT(*)
    FROM release_check
    WHERE override_schema_version IS NULL
    GROUP BY release_id
    HAVING COUNT(*) > 1
) tmp;
```

has 544950 rows. Setting `override_schema_version` to empty string would cause unique violation errors.

```sql
SELECT DISTINCT(collection_id)
FROM release_with_collection
WHERE id IN (
    SELECT release_id
    FROM release_check
    WHERE override_schema_version IS NULL
    GROUP BY release_id
    HAVING COUNT(*) > 1
);
```

The collections with releases with duplicate release checks are: 14, 15, 16, 19, 20, 32, 35, 41, 43, 44, 62, 76.

```sql
SELECT id, source_id, data_version, sample, transform_type, check_data FROM collection
WHERE id IN (14, 15, 16, 19, 20, 32, 35, 41, 43, 44, 62, 76);
```

```
 id |   source_id    |    data_version     | sample |   transform_type   | check_data 
----+----------------+---------------------+--------+--------------------+------------
 14 | honduras-oncae | 2019-02-14 10:00:12 | f      |                    | t
 15 | buenos-aires   | 2019-02-18 10:00:12 | f      |                    | t
 16 | colombia       | 2019-02-19 11:08:24 | f      |                    | t
 19 | colombia       | 2019-02-21 17:07:08 | f      |                    | t
 35 | colombia       | 2019-02-26 12:15:28 | f      | upgrade-1-0-to-1-1 | f
 43 | colombia       | 2019-02-27 22:14:28 | f      |                    | t
 76 | mexico-gacme   | 2019-03-11 14:43:12 | f      |                    | t
 32 | colombia       | 2019-02-25 11:37:53 | f      | upgrade-1-0-to-1-1 | f
 41 | colombia       | 2019-02-27 16:17:47 | f      | upgrade-1-0-to-1-1 | f
 44 | colombia       | 2019-02-27 22:14:28 | f      | upgrade-1-0-to-1-1 | f
 20 | colombia       | 2019-02-21 17:07:08 | f      | upgrade-1-0-to-1-1 | f
 62 | colombia       | 2019-03-08 14:25:38 | f      | upgrade-1-0-to-1-1 | f
```

I don't think we need the check results from all those upgraded collection (we typically only check source collections); indeed, `check_data` is false for these, so maybe they were run manually?

For the others, it might be quickest to delete the check results, and then run the checks worker?

**Comment by odscjames** [comment]
> For the others, it might be quickest to delete the check results, and then run the checks worker?

Checks will re-run if deleted, so the quickest pass here is to delete all checks for relevant collections and just let the normal background process re-do them. We should check with analysts they are not in actual use first.

Tho, we should check with analysts anyway - it may be we can just delete these collections totally anyway. They are all old, and I know some some of those source_ids have newer downloads.

**Comment by odscjames** [comment]
> it may be we can just delete these collections totally anyway. 

I have checked and got permission. I will mark them deleted and start a delete process now; not sure how long it will take to run.

**Comment by jpmckinney** [comment]
```sql
SELECT release_id
FROM release_check
WHERE override_schema_version IS NULL
GROUP BY release_id
HAVING COUNT(*) > 1
```

now returns 0 rows, so I'll add it to #249.

**Comment by jpmckinney** [comment]
Follow-up issues are #251 and https://github.com/open-contracting/kingfisher-scrape/issues/237

The rest will be resolved by the PRs #248 #249 #250
